File: src/main/java/platform/ecommerce/domain/common/vo/Money.java
package platform.ecommerce.domain.common.vo;

import jakarta.persistence.Column;
import jakarta.persistence.Embeddable;
import lombok.AccessLevel;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.math.RoundingMode;

/**
 * Value Object representing monetary amounts.
 * Immutable and provides arithmetic operations.
 */
@Embeddable
@Getter
@EqualsAndHashCode
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Money {

    public static final Money ZERO = new Money(BigDecimal.ZERO);

    @Column(name = "amount", precision = 12, scale = 2)
    private BigDecimal amount;

    private Money(BigDecimal amount) {
        this.amount = amount.setScale(2, RoundingMode.HALF_UP);
    }

    public static Money of(BigDecimal amount) {
        if (amount == null) {
            throw new IllegalArgumentException("Amount cannot be null");
        }
        return new Money(amount);
    }

    public static Money of(long amount) {
        return new Money(BigDecimal.valueOf(amount));
    }

    public static Money of(String amount) {
        if (amount == null || amount.isBlank()) {
            throw new IllegalArgumentException("Amount cannot be null or empty");
        }
        return new Money(new BigDecimal(amount));
    }

    public Money add(Money other) {
        return new Money(this.amount.add(other.amount));
    }

    public Money subtract(Money other) {
        return new Money(this.amount.subtract(other.amount));
    }

    public Money multiply(int multiplier) {
        return new Money(this.amount.multiply(BigDecimal.valueOf(multiplier)));
    }

    public Money multiply(BigDecimal multiplier) {
        return new Money(this.amount.multiply(multiplier));
    }

    public Money percentage(BigDecimal percentage) {
        return new Money(this.amount.multiply(percentage).divide(BigDecimal.valueOf(100), 2, RoundingMode.HALF_UP));
    }

    public boolean isGreaterThan(Money other) {
        return this.amount.compareTo(other.amount) > 0;
    }

    public boolean isGreaterThanOrEqual(Money other) {
        return this.amount.compareTo(other.amount) >= 0;
    }

    public boolean isLessThan(Money other) {
        return this.amount.compareTo(other.amount) < 0;
    }

    public boolean isPositive() {
        return this.amount.compareTo(BigDecimal.ZERO) > 0;
    }

    public boolean isZero() {
        return this.amount.compareTo(BigDecimal.ZERO) == 0;
    }

    public boolean isNegative() {
        return this.amount.compareTo(BigDecimal.ZERO) < 0;
    }

    @Override
    public String toString() {
        return amount.toString();
    }
}

File: src/main/kotlin/platform/ecommerce/domain/vo/Money.kt
package platform.ecommerce.domain.vo

import jakarta.persistence.Embeddable
import java.math.BigDecimal

@Embeddable
data class Money(
    val amount: BigDecimal
) {
    init {
        require(amount >= BigDecimal.ZERO) {
            "Amount must be greater than or equal to zero."
        }
    }

    companion object {
        val ZERO = Money(BigDecimal.ZERO)

        fun of(amount: Double): Money {
            return Money(BigDecimal(amount))
        }
    }

    fun plus(other: Money): Money {
        return Money(this.amount + other.amount)
    }

    fun minus(other: Money): Money {
        val result = this.amount - other.amount
        require(result >= BigDecimal.ZERO) {
            "Resulting amount must be greater than or equal to zero."
        }
        return Money(result)
    }

    fun times(factor: Int): Money {
        require(factor >= 0) {
            "Factor must be greater than or equal to zero."
        }
        return Money(this.amount * BigDecimal.valueOf(factor.toLong()))
    }
}
